### PR TITLE
Views refactoring + one bug fix

### PR DESCRIPTION
**Pull request by chrismedrela**
- Bug fix in `all_trainings`: The number of trainees who became SWC or DC instructor was overestimated due to a bug in a query.
- Switching to class-based views in case of simple `ListView`s and some simple `DetailView`s.
- Adopting new name convention (`AMYWhateverView`) for our generic views.
- Moving generic views from `views.py` to `base_views.py`.

@pbanaszkiewicz what do you think about these changes?

### CODE REVIEW

File: workshops/base_views.py [hunk 1]
```diff
@@ -0,0 +1,208 @@
...
+            return HttpResponseRedirect(success_url)
+        except ProtectedError as e:
+            return failed_to_delete(self.request, self.object,
+                                    e.protected_objects)
+
+    def get(self, request, *args, **kwargs):
+        return self.http_method_not_allowed(request, *args, **kwargs)
+
+
+class FormViewContext(FormView):
```
Comment (narayanaditya95) [inline workshops/base_views.py]: `FormViewContext` -> `AMYFormView`?

**Comment by narayanaditya95** [comment]
@chrismedrela Can you make the corresponding changes in `pydata/views.py` in this PR?

**Comment by chrismedrela** [comment]
@narayanaditya95 what changes do you mean?

**Comment by pbanaszkiewicz** [comment]
@chrismedrela:
> @narayanaditya95 what changes do you mean?

Aditya suggested renaming `FormViewContext` to `AMYFormView` (seems reasonable). He also asked if we should make `views` a package, but I think it's a little bit too much work - we could do this for v1.10, though.

**Comment by chrismedrela** [comment]
> Aditya suggested renaming FormViewContext to AMYFormView (seems reasonable).
OK

So now it's ready. I've added a few more commits, mainly converting some detailed view to class-based views.

**Comment by pbanaszkiewicz** [comment]
Thanks @chrismedrela

**Pull request merged**